Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-UpgradedServiceFabricApplication.ps1

```diff
@@ -359,24 +359,105 @@ function Publish-UpgradedServiceFabricApplication
             return
         }
 
-        $upgradeStatus = Wait-ServiceFabricApplicationUpgradeAction -ApplicationName $ApplicationName
+        $unhealthyEvaluationMessageFunction = {
+            param(
+                [object]$UnhealthyEvaluations,
 
-        if ($UnregisterUnusedVersions)
-        {
-            Write-Host (Get-VstsLocString -Key SFSDK_UnregisterUnusedVersions)
-            foreach ($registeredAppType in Get-ServiceFabricApplicationTypeAction -ApplicationTypeName $names.ApplicationTypeName)
+                [string]$Indentation
+            )
+
+            if ($UnhealthyEvaluations -eq $null) {
+            return ""
+            }
+
+            $unhealthyEvaluationsKind = ($UnhealthyEvaluations.kind | Out-String).Trim()
+            $indentatedErrorString = ""
+            $indentatedErrorString += $Indentation + $UnhealthyEvaluations.Description + "`n"
+            foreach ($UnhealthyEvaluation in $UnhealthyEvaluations.UnhealthyEvaluations)
             {
-                try
+                # see if indentation needs to be increased. based on the type of evaluation.
+                $unhealthyEvaluationKind = ($UnhealthyEvaluation.kind | Out-String).Trim()
+                $newIndentation = $Indentation
+                if (!($unhealthyEvaluationsKind -eq ($unhealthyEvaluationKind + "s")))
                 {
-                    Unregister-ServiceFabricApplicationTypeAction -ApplicationTypeName $($registeredAppType.ApplicationTypeName) -ApplicationTypeVersion $($registeredAppType.ApplicationTypeVersion) -TimeoutSec $UnregisterPackageTimeoutSec
+                    $newIndentation += "`t"
                 }
-                catch
+
+                $indentatedErrorString += (& $unhealthyEvaluationMessageFunction -UnhealthyEvaluations $UnhealthyEvaluation -Indentation $newIndentation) + "`n"
+            }
+
+            if ($UnhealthyEvaluations.UnhealthyEvent)
+            {
+                $indentatedErrorString += $Indentation + $UnhealthyEvaluations.UnhealthyEvent.HealthInformation.Description + "`n"
+            }
+
+            return $indentatedErrorString
+        }
+
+        $DomainUpgradeStatusFormatterFunction = {
+            param(
+                [object]$UpgradeDomainsStatus
+            )
+            if ($UpgradeDomainsStatus -eq $null)
+            {
+                return ""
+            }
+
+            $upgradeDomainStatusString = ([String]($UpgradeDomainsStatus)).Trim()
+            return $upgradeDomainStatusString
+        }
+
+        $upgradeStatusFetcher = {
+            param(
+                [object]$LastUpgradeStatus
+            )
+            $upgradeStatus = Get-ServiceFabricApplicationUpgrade -ApplicationName $ApplicationName
+            Write-Host (Get-VstsLocString -Key SFSDK_CurrentUpgradeState) $upgradeStatus.UpgradeState
+
+            $currentDomainWiseUpgradeStatus = & $DomainUpgradeStatusFormatterFunction -UpgradeDomainsStatus $upgradeStatus.UpgradeDomainsStatus
```

Comment:
move this logic to a function. this will keep the code clean

Reply:
And keep that under try-catch so that any failure does not fail task

Reply:
The retry function already keeps it under try catch !